### ISSUE TITLE
Tree corruption when deleting items one-by-one while adding new items

### ISSUE THREAD

**Issue by velis74** [django-mptt/django-mptt#417]
As the title says: I tried to duplicate part of structure into a different parent node while at the same time deleting original nodes. This lead to tree corruption.

The (much) longer explanation with code that yields corruption:

Sometimes I have to copy / move / delete items in a selected subtree.
So I build myself a simple recursive structure that remembers all items. Then I copy these items to new parents while at the same time deleting previous nodes.

Here's the entire class that does that:

    class Children(object):
        def __init__(self, parent: SiteTree, deletable=True):
            self.sitetree = parent if deletable else None
            self.site = parent.site
            # Children are always not deletable because deleting parent will also delete children
            self.children = [Children(child, False) for child in parent.get_children()]

        def rebase(self, newp):
            for child in self.children:
                site = child.site
                np = [SiteTree.objects.create(site=site, parent=p) for p in newp]
                child.rebase(np)
            if self.sitetree:
                self.sitetree.delete()

Please note the comment in `__init__`: if I build the structure such that I pass deletable to Children being created as in 

    self.children = [Children(child, deletable) for child in parent.get_children()]

Then `rebase` method will attempt to delete each child immediately after it processes its children.
In a simple structure, such as

    master
    - base1
        - parent 1
            - item 1
                - subitem 1
            - item 2
                - subitem 2
    - base 2

When trying to move `parent 1` from `base1` to `base2` this results in corrupted structure immediately after deleting `subitem 2` (having already made a copy of structure for all items & subitems @ `base2`).

If however I only delete the original `parent1` node, the structure is left operational.

Structure corruption is such that rebuild() is successful in reestablishing it. Without rebuild, however, any subsequent operations have undetermined results.

**Comment by GlennV** [comment]
This problem also occured to us while doing something very similar, but manually. A user made some new nodes and deleted some old ones. This lead to tree corruption so that get_ancestors(include_self=True) became very undetermined.

However, I don't have the necessary data to reproduce the issue. A rebuild() fixed the problem, but as we  have a multi-tenant database, tree corruption leads to dangerous data integrity issues. Any idea what is causing this corruption?

**Comment by MichelML** [comment]
same issue on our end here :+1:

**Comment by matthiask** [comment]
Yes, that's still not supported. You either have to use the `delay_mptt_updates` context manager or `refresh_from_db` your models' mptt attributes before saving.

This is one of the very sharp edges of django-mptt and it's not probable that this will ever be fixed. django-mptt isn't optimized for writes but for reads. 

For more about that see https://406.ch/writing/django-tree-queries/

**Comment by matthiask** [comment]
Duplicate of #257 — the root cause is saving model instances whose MPTT fields are out of sync with the database. Use `delay_mptt_updates()` around bulk operations, or call `instance.refresh_from_db()` before saving. See #257 for the full discussion.
